// deso-protocol/src/transactions/derived-keys.js
import { bytesToHex, hexToBytes } from '@noble/hashes/utils';
import { TransactionExtraDataKV, TransactionMetadataAuthorizeDerivedKey, bs58PublicKeyToCompressedBytes, encodeUTF8ToBytes, } from '../identity/index.js';
import { constructBalanceModelTx, handleSignAndSubmit } from '../internal.js';
export const authorizeDerivedKey = (params, options) => {
    return handleSignAndSubmit('api/v0/authorize-derived-key', params, {
        ...options,
        constructionFunction: constructAuthorizeDerivedKey,
    });
};
export const constructAuthorizeDerivedKey = (params) => {
    const metadata = new TransactionMetadataAuthorizeDerivedKey();
    metadata.accessSignature = hexToBytes(params.AccessSignature || '');
    metadata.derivedPublicKey = bs58PublicKeyToCompressedBytes(params.DerivedPublicKeyBase58Check);
    metadata.expirationBlock = params.ExpirationBlock;
    metadata.operationType = params.DeleteKey ? 0 : 1;
    const consensusExtraDataKVs = [];
    if (params.DerivedKeySignature) {
        consensusExtraDataKVs.push(new TransactionExtraDataKV(encodeUTF8ToBytes('DerivedPublicKey'), bs58PublicKeyToCompressedBytes(params.DerivedPublicKeyBase58Check)));
    }
    if (params.TransactionSpendingLimitHex) {
        const transactionSpendingLimitBuf = hexToBytes(params.TransactionSpendingLimitHex);
        if (transactionSpendingLimitBuf.length) {
            consensusExtraDataKVs.push(new TransactionExtraDataKV(encodeUTF8ToBytes('TransactionSpendingLimit'), transactionSpendingLimitBuf));
        }
    }
    if (params.Memo || params.AppName) {
        const memo = params.Memo || params.AppName;
        consensusExtraDataKVs.push(new TransactionExtraDataKV(encodeUTF8ToBytes('DerivedKeyMemo'), encodeUTF8ToBytes(bytesToHex(encodeUTF8ToBytes(memo)))));
    }
    return constructBalanceModelTx(params.OwnerPublicKeyBase58Check, metadata, {
        ConsensusExtraDataKVs: consensusExtraDataKVs,
        ExtraData: params.ExtraData,
        MinFeeRateNanosPerKB: params.MinFeeRateNanosPerKB,
        TransactionFees: params.TransactionFees,
    });
};
