// deso-protocol/src/identity/types.js
export var NOTIFICATION_EVENTS;
(function (NOTIFICATION_EVENTS) {
    /**
     * This event is fired when the consuming app initially subscribes to identity.
     */
    NOTIFICATION_EVENTS["SUBSCRIBE"] = "SUBSCRIBE";
    /**
     * This is an intermediate event fired AFTER the user completes an identity flow
     * that requires a derived key authorization. This event is fired BEFORE the
     * request to authorize the derived key is made.
     */
    NOTIFICATION_EVENTS["AUTHORIZE_DERIVED_KEY_START"] = "AUTHORIZE_DERIVED_KEY_START";
    /**
     * This is an intermediate event fired AFTER the user completes an identity
     * flow that requires a derived key authorization. This event is fired AFTER
     * the request to authorize the derived key is made.
     */
    NOTIFICATION_EVENTS["AUTHORIZE_DERIVED_KEY_END"] = "AUTHORIZE_DERIVED_KEY_END";
    /**
     * This event is fired if the request to authorize a derived key fails.
     */
    NOTIFICATION_EVENTS["AUTHORIZE_DERIVED_KEY_FAIL"] = "AUTHORIZE_DERIVED_KEY_FAIL";
    /**
     * This event is fired when the user opens the permissions approval popup.
     */
    NOTIFICATION_EVENTS["REQUEST_PERMISSIONS_START"] = "REQUEST_PERMISSIONS_START";
    /**
     * This event is fired when the user completes approving permissions, and
     * comes AFTER the intermediate AUTHORIZE_DERIVED_KEY events.
     */
    NOTIFICATION_EVENTS["REQUEST_PERMISSIONS_END"] = "REQUEST_PERMISSIONS_END";
    /**
     * This event is fired when the user opens the login popup.
     */
    NOTIFICATION_EVENTS["LOGIN_START"] = "LOGIN_START";
    /**
     * This event is fired when the user completes logging in, and
     * comes AFTER the intermediate AUTHORIZE_DERIVED_KEY events.
     */
    NOTIFICATION_EVENTS["LOGIN_END"] = "LOGIN_END";
    /**
     * This event is fired when the user opens the logout popup.
     */
    NOTIFICATION_EVENTS["LOGOUT_START"] = "LOGOUT_START";
    /**
     * This event is fired when the user completes logging out, and
     * comes AFTER the intermediate AUTHORIZE_DERIVED_KEY events.
     */
    NOTIFICATION_EVENTS["LOGOUT_END"] = "LOGOUT_END";
    /**
     * This event is fired when the user opens the get deso popup.
     */
    NOTIFICATION_EVENTS["GET_FREE_DESO_START"] = "GET_FREE_DESO_START";
    /**
     * This event is fired when the user completes the get deso flow, and comes
     * AFTER the intermediate AUTHORIZE_DERIVED_KEY events.
     */
    NOTIFICATION_EVENTS["GET_FREE_DESO_END"] = "GET_FREE_DESO_END";
    /**
     * This event is fired when the user opens the verify phone number popup.
     */
    NOTIFICATION_EVENTS["VERIFY_PHONE_NUMBER_START"] = "VERIFY_PHONE_NUMBER_START";
    /**
     * This event is fired when the user completes the verify phone number flow,
     * and comes AFTER the intermediate AUTHORIZE_DERIVED_KEY events.
     */
    NOTIFICATION_EVENTS["VERIFY_PHONE_NUMBER_END"] = "VERIFY_PHONE_NUMBER_END";
    /**
     * This event is fired when the consuming app switches the active user.
     */
    NOTIFICATION_EVENTS["CHANGE_ACTIVE_USER"] = "CHANGE_ACTIVE_USER";
})(NOTIFICATION_EVENTS || (NOTIFICATION_EVENTS = {}));
