// deso-protocol/src/identity/api.js
function buildOptions(customOptions = {}) {
    const headers = customOptions.headers ?? {};
    delete customOptions.headers;
    return {
        headers,
        ...customOptions,
    };
}
export class APIError {
    constructor(message, status) {
        Object.defineProperty(this, "message", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "status", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.message = message;
        this.status = status;
    }
    toString() {
        return this.message;
    }
}
const wrappedFetch = (url, options) => {
    return fetch(url, options).then((res) => {
        if (!res.ok) {
            return res.json().then((json) => {
                throw new APIError(json.error, res.status);
            });
        }
        return res.text().then((rawBodyText) => {
            if (rawBodyText.length === 0) {
                // In the case where the response body is empty, it will fail to parse
                // as JSON so we just handle it as a special case.
                return null;
            }
            else {
                return JSON.parse(rawBodyText);
            }
        });
    });
};
export const api = {
    post(url, data) {
        return wrappedFetch(url, buildOptions({
            method: 'POST',
            body: JSON.stringify(data),
            headers: {
                'Content-Type': 'application/json',
            },
        }));
    },
    get(url) {
        return wrappedFetch(url, buildOptions({ method: 'GET' }));
    },
};
